src/solvers/refinement/string_constraint.h

```diff
@@ -0,0 +1,192 @@
+
+  inline const exprt &premise() const
+  {
+    return op0();
+  }
+
+  inline const exprt &body() const
+  {
+    return op1();
+  }
+
+  inline const exprt &univ_var() const
+  {
+    return op2();
+  }
+
+  inline const exprt &upper_bound() const
+  {
+    return op3();
+  }
+
+  inline const exprt &lower_bound() const
+  {
+    return operands()[4];
+  }
+
+  // Trivial constraint
+  string_constraintt() : exprt(ID_string_constraint)
+  {
+    assert(false); // string constraints should not be initialized directly
+    copy_to_operands(true_exprt(), true_exprt());
+  }
+
+  string_constraintt(
+    const symbol_exprt &univ,
+    const exprt &bound_inf,
+    const exprt &bound_sup,
+    const exprt &prem,
+    const exprt &body)
+    : exprt(ID_string_constraint)
+  {
+    copy_to_operands(prem, body);
+    copy_to_operands(univ, bound_sup, bound_inf);
+  };
+
+  // Default bound inferior is 0
+  string_constraintt(
+    const symbol_exprt &univ,
+    const exprt &bound_sup,
+    const exprt &prem,
+    const exprt &body)
+    : string_constraintt(univ, refined_string_typet::index_zero(),
+                         bound_sup, prem, body)
+  {};
+
+  // Default premise is true
+  string_constraintt
+  (const symbol_exprt &univ, const exprt &bound_sup, const exprt &body)
+    : string_constraintt
+      (univ, refined_string_typet::index_zero(), bound_sup, true_exprt(), body)
+  {};
+
+  bool is_simple() const { return (operands().size()==2); }
+  bool is_univ_quant() const { return (operands().size()==5); }
+  bool is_not_contains() const { return false; }
+
+  inline symbol_exprt get_univ_var() const
+  { return to_symbol_expr(univ_var()); }
+
+  inline exprt univ_within_bounds() const
+  {
+    return and_exprt
+      (binary_relation_exprt(lower_bound(), ID_le, get_univ_var()),
+       binary_relation_exprt(upper_bound(), ID_gt, get_univ_var()));
+  }
+};
+
+extern inline const string_constraintt &to_string_constraint(const exprt &expr)
+{
+  assert(expr.id()==ID_string_constraint && expr.operands().size()==5);
+  return static_cast<const string_constraintt &>(expr);
+}
+
+extern inline string_constraintt &to_string_constraint(exprt &expr)
+{
+  assert(expr.id()==ID_string_constraint && expr.operands().size()==5);
+  return static_cast<string_constraintt &>(expr);
+}
+
+class string_not_contains_constraintt: public exprt
+{
+public:
+  // string_not contains_constraintt are formula of the form:
+  // forall x in [lb,ub[. p(x) => exists y in [lb,ub[. s1[x+y] != s2[y]
+
+  string_not_contains_constraintt(
+    exprt univ_lower_bound,
+    exprt univ_bound_sup,
+    exprt premise,
+    exprt exists_bound_inf,
+    exprt exists_bound_sup,
+    exprt s0,
+    exprt s1)
+    :exprt(ID_string_not_contains_constraint)
+  {
+    copy_to_operands(univ_lower_bound, univ_bound_sup, premise);
+    copy_to_operands(exists_bound_inf, exists_bound_sup, s0);
+    copy_to_operands(s1);
+  };
+
+  bool is_not_contains() const { return true; }
```

Comment:
yes you are right, it's redundant and I removed it: https://github.com/diffblue/cbmc/pull/374/commits/3cb5cd4253a6af4d82a0f089d8682df1dc5073ac